sw/device/lib/dif/dif_gpio.c

```diff
@@ -70,96 +70,112 @@ void (*abort_if_invalid_fptr)(dif_gpio_t const *const) = abort_if_invalid;
 }
 #ifdef GTEST_TEST
-bool (*read_register_bit_fptr)(dif_gpio_t const *const gpio,
-                               uint32_t const offset,
-                               uint32_t const index) = read_register_bit;
+bool (*read_register_bit_fptr)(const dif_gpio_t *const gpio,
+                               const uint32_t offset,
+                               const uint32_t index) = read_register_bit;
 #define read_register_bit read_register_bit_fptr
 #endif
 
-static inline void write_register(dif_gpio_t const *const gpio,
-                                  uint32_t const offset, uint32_t const val) {
+static inline void write_register(const dif_gpio_t *const gpio,
+                                  const uint32_t offset, const uint32_t val) {
   abort_if_invalid(gpio);
   reg32_write(gpio->base_addr, offset, val);
 }
 #ifdef GTEST_TEST
-void (*write_register_fptr)(dif_gpio_t const *const gpio, uint32_t const offset,
-                            uint32_t const val) = write_register;
+void (*write_register_fptr)(const dif_gpio_t *const gpio, const uint32_t offset,
+                            const uint32_t val) = write_register;
 #define write_register write_register_fptr
 #endif
 
-static inline void write_register_masked_access(dif_gpio_t const *const gpio,
-                                                uint32_t const lowerOffset,
-                                                uint32_t const upperOffset,
-                                                uint32_t const mask,
-                                                uint32_t const val) {
+/**
+ * Perform a masked (thread-safe) write to upper or lower half of a register.
+ *
+ * GPIO device provides masked (thread-safe) writes to some of the registers.
+ * This allows software to modify half of the bits at a time without requiring
+ * a read-modify-write. Note that depending on the value of the |mask|, this
+ * function may perform two writes.
+ *
+ * @param gpio GPIO instance
+ * @param lowerOffset Offset of the masked access register to modify the lower
+ * half of a register.
+ * @param upperOffset Offset of the masked access register to modify the upper
+ * half of a register.
+ * @param mask Mask that identifies the bits to write to.
+ * @param val Value to write
+ */
+static inline void write_register_masked_access(const dif_gpio_t *const gpio,
+                                                const uint32_t lowerOffset,
+                                                const uint32_t upperOffset,
+                                                const uint32_t mask,
+                                                const uint32_t val) {
   // Check here once and use |reg32_write()| instead of |write_register()|,
   // which may result in two calls to |abort_if_invalid()|.
   abort_if_invalid(gpio);
-  uint32_t const lower_mask = mask & 0x0000FFFFU;
-  uint32_t const upper_mask = mask & 0xFFFF0000U;
-  if (lower_mask) {
+  const uint32_t lower_mask = mask & 0x0000FFFFU;
```

Comment:
You are correct, but this function uses masked access to write to anywhere in the register. The `mask` parameter is over the entire register and the actual write uses masked access to write to `lowerOffset` or `upperOffset` depending on the value of `mask`.

That being said, I wonder if we need to reconsider my approach here. Masked access guarantees thread-safe writes to upper or lower half of a register. This function will do a thread-safe write if the caller is only writing to the upper or lower half but if the caller is writing to both halves, we will have two writes.

Reply:
Ohh I see, thank you for the explanation. I think this function and variable naming is slightly confusing, because you use variable and parameter names `upper, lower, lowerOffset, upperOffset`. These sound similar, but mean completely different things. I think this function might be a bit too clever, please consider splitting it in two separate functions:
```
void gpio_set_masked_upper(const dif_gpio_t *const gpio, const uint32_t mask, const uint32_t val)
void gpio_set_masked_lower(const dif_gpio_t *const gpio, const uint32_t mask, const uint32_t val)
```